File: leetcode_python/Heap/least-number-of-unique-integers-after-k-removals.py
"""

1481. Least Number of Unique Integers after K Removals
Medium

Given an array of integers arr and an integer k. Find the least number of unique integers after removing exactly k elements.

 

Example 1:

Input: arr = [5,5,4], k = 1
Output: 1
Explanation: Remove the single 4, only 5 is left.
Example 2:
Input: arr = [4,3,1,1,3,3,2], k = 3
Output: 2
Explanation: Remove 4, 2 and either one of the two 1s or three 3s. 1 and 3 will be left.
 

Constraints:

1 <= arr.length <= 10^5
1 <= arr[i] <= 10^9
0 <= k <= arr.length

"""

# V0
# IDEA : Counter
from collections import Counter
class Solution:
    def findLeastNumOfUniqueInts(self, arr, k):
        # edge case
        if not arr:
            return 0
        cnt = dict(Counter(arr))
        cnt_sorted = sorted(cnt.items(), key = lambda x : x[1])
        #print ("cnt_sorted = " + str(cnt_sorted))
        removed = 0
        for key, freq in cnt_sorted:
            """
            NOTE !!!
                -> we need to remove exactly k elements and make remain unique integers as less as possible
                -> since we ALREADY sort num_counter,
                -> so the elements NOW are ordering with their count
                    -> so we need to remove ALL element while k still > 0
                    -> so k -= freq, since for element key, there are freq count for it in arr
            """
            if freq <= k:
                k -= freq
                removed += 1

        return len(cnt.keys()) - removed

# V0
# IDEA : Counter
class Solution:
    def findLeastNumOfUniqueInts(self, arr, k):
        num_counter = Counter(arr)
        remove_ele = []
        _num_counter = sorted(num_counter.items(), key = lambda x : x[1])
        for key, freq in _num_counter:
            """
            NOTE !!!
                -> we need to remove exactly k elements and make remain unique integers as less as possible
                -> since we ALREADY sort num_counter,
                -> so the elements NOW are ordering with their count
                    -> so we need to remove ALL element while k still > 0
                    -> so k -= freq, since for element key, there are freq count for it in arr
            """
            if freq <= k:
                k -= freq
                remove_ele.append(key)
        return len(num_counter) - len(remove_ele)

# V0'
# IDEA : Counter
class Solution:
    def findLeastNumOfUniqueInts(self, arr, k):
        num_counter = Counter(arr)
        remove_ele = []
        for key, freq in sorted(num_counter.items(), key = lambda x : x[1]):
            if freq <= k:
                k -= freq
                remove_ele.append(key)
        return len(num_counter) - len(remove_ele)

# V0''
# IDEA : Counter + heapq
class Solution(object):
    def findLeastNumOfUniqueInts(self, arr, k):
            # use counter, and heap (priority queue)
            from collections import Counter
            import heapq
            h = []
            for key, val in Counter(arr).items():
                heapq.heappush(h,(val,key))

            while k > 0:
                item = heapq.heappop(h)    
                if item[0] != 1:
                    heapq.heappush(h, (item[0]-1, item[1]))      
                k -=1

            return len(h)

# V1
# IDEA : Counter
# https://leetcode.com/problems/least-number-of-unique-integers-after-k-removals/discuss/979896/Python-Solution
class Solution:
    def findLeastNumOfUniqueInts(self, arr, k):
        num_counter = Counter(arr)
        remove_ele = []
        for key, freq in sorted(num_counter.items(), key = lambda x : x[1]):
            if freq <= k:
                k -= freq
                remove_ele.append(key)
        return len(num_counter) - len(remove_ele)

# V1'
# IDEA : Counter
# https://leetcode.com/problems/least-number-of-unique-integers-after-k-removals/discuss/686293/Python-solution-with-Counter
class Solution:
    def findLeastNumOfUniqueInts(self, arr, k):
        from collections import Counter
        count = sorted(Counter(arr).items(), key=lambda x: x[1])  # sort dictionary by value increasing and get it as tuple
        removed = 0  # number of removed items
        for key, val in count: 
            if k >= val:  
                k -= val
                removed += 1
            else:
                break
        return len(count)-removed  # number of remained elements

# V1''
# IDEA : Counter
from collections import Counter
class Solution:
    def findLeastNumOfUniqueInts(self, arr, k):
        """
        Given an array of integers arr and number of removals k, this
        program uses the Python Counter to help it determine the least
        number of unique integers in arr after k removals.

        :param arr: array of integers
        :type arr: list[int]
        :param k: number of integer removals from arr
        :type k: int
        :return: least number of unique integers in arr after k removals
        :rtype: int
        """
        """
        Remove the integers with the lowest counts to get the least number
        of unique integers in arr
        """
        counts = Counter( arr ).most_common()
        unique = len( counts )
        while k > 0 and unique > 0:
            k -= counts[unique - 1][1]
            if k >= 0:
                unique -= 1
        return unique

# V1'''
# IDEA : Counter + heapq
# https://leetcode.com/problems/least-number-of-unique-integers-after-k-removals/discuss/686429/Python-heaps
from collections import Counter
from heapq import *
class Solution(object):
    def findLeastNumOfUniqueInts(self, arr, k):
        c = Counter(arr)
        heap = []
        for key, value in c.items():
            heappush(heap, (value, key))
        while heap and k > 0:
            value, key = heappop(heap)
            if value > 1:
                heappush(heap, (value-1, key))
            k -= 1
        return len(heap)

# V1''''
# IDEA : Counter + heapq
# https://leetcode.com/problems/least-number-of-unique-integers-after-k-removals/discuss/688829/Python-heap-solution
class Solution(object):
    def findLeastNumOfUniqueInts(self, arr, k):
        heap = []
        count = collections.Counter(arr)
        for key in count:
            heapq.heappush(heap, (count[key], key))
        
        while(heap and k):
            count, key = heapq.heappop(heap)
            if k >= count:
                k -= count
            else:
                return len(heap) + 1
        return len(heap)

# V1'''''
# IDEA : Counter + heapq
# https://leetcode.com/problems/least-number-of-unique-integers-after-k-removals/discuss/704179/python-solution%3A-Counter-and-Priority-Queue
# IDEA
# -> Count the occurence of each number.
# -> We want to delete the number with lowest occurence thus we can use minimum steps to reduce the total unique numbers in the list. For example,[4,3,1,1,3,3,2]. The Counter of this array will be: {3:3, 1:2, 4:1, 2:1}. Given k = 3, the greedy approach is to delete 2 and 4 first because both of them are appearing once. We need an ordering data structure to give us the lowest occurence of number each time. As you may know, Priority Queue comes to play
# -> Use heap to build PQ for the counter. We store each member as a tuple: (count, number) Python heap module will sort it based on the first member of the tuple.
# -> loop through k times to pop member out of heap and check if we need to push it back
class Solution(object):
    def findLeastNumOfUniqueInts(self, arr, k):
            # use counter, and heap (priority queue)
            from collections import Counter
            import heapq
            h = []
            for key, val in Counter(arr).items():
                heapq.heappush(h,(val,key))

            while k > 0:
                item = heapq.heappop(h)    
                if item[0] != 1:
                    heapq.heappush(h, (item[0]-1, item[1]))      
                k -=1

            return len(h)

# V1'''''''
# IDEA : Counter + heapq
# https://leetcode.com/problems/least-number-of-unique-integers-after-k-removals/discuss/1542356/Python-MinHeap-Solution
class Solution:
    def findLeastNumOfUniqueInts(self, arr, k):
        counter = collections.Counter(arr)
        minHeap = []
        for key, val in counter.items():
            heapq.heappush(minHeap, val)
        
        while k:
            minHeap[0] -= 1
            if minHeap[0] == 0:
                heapq.heappop(minHeap)
            k -= 1
        
        return len(minHeap)

# V1'''''''
# IDEA : GREEDY
# https://zxi.mytechroad.com/blog/hashtable/leetcode-1481-least-number-of-unique-integers-after-k-removals/
# C++
# class Solution {
# public:
#   int findLeastNumOfUniqueInts(vector<int>& arr, int k) {    
#     unordered_map<int, int> c;
#     for (int x : arr) ++c[x];
#     vector<int> m; // freq
#     for (const auto [x, f] : c)
#       m.push_back(f);
#     sort(begin(m), end(m));
#     int ans = m.size();    
#     int i = 0;
#     while (k--) {
#       if (--m[i] == 0) {
#         ++i;
#         --ans;
#       }
#     }
#     return ans;
#   }
# };

# V2

File: leetcode_java/src/main/java/LeetCodeJava/Heap/LeastNumberOfUniqueIntegersAfterKRemovals.java
package LeetCodeJava.Heap;

// https://leetcode.com/problems/least-number-of-unique-integers-after-k-removals/description

import java.util.*;

/**
 *  1481. Least Number of Unique Integers after K Removals
 * Solved
 * Medium
 * Topics
 * premium lock icon
 * Companies
 * Hint
 * Given an array of integers arr and an integer k. Find the least number of unique integers after removing exactly k elements.
 *
 *
 *
 * Example 1:
 *
 * Input: arr = [5,5,4], k = 1
 * Output: 1
 * Explanation: Remove the single 4, only 5 is left.
 * Example 2:
 * Input: arr = [4,3,1,1,3,3,2], k = 3
 * Output: 2
 * Explanation: Remove 4, 2 and either one of the two 1s or three 3s. 1 and 3 will be left.
 *
 *
 * Constraints:
 *
 * 1 <= arr.length <= 10^5
 * 1 <= arr[i] <= 10^9
 * 0 <= k <= arr.length
 *
 */
public class LeastNumberOfUniqueIntegersAfterKRemovals {

    // V0
//    public int findLeastNumOfUniqueInts(int[] arr, int k) {
//
//    }

    // V1-1
    // IDEA: Sorting the Frequencies
    // https://leetcode.com/problems/least-number-of-unique-integers-after-k-removals/
    public int findLeastNumOfUniqueInts_1_1(int[] arr, int k) {
        // Map to track the frequencies of elements
        Map<Integer, Integer> map = new HashMap<>();
        for (int i : arr) {
            map.put(i, map.getOrDefault(i, 0) + 1);
        }

        // List to track all the frequencies
        List<Integer> frequencies = new ArrayList<>(map.values());

        // Sorting the frequencies
        Collections.sort(frequencies);

        // Tracking the number of elements removed
        int elementsRemoved = 0;

        for (int i = 0; i < frequencies.size(); i++) {
            // Removing frequencies[i] elements, which equates to
            // removing one unique element
            elementsRemoved += frequencies.get(i);

            // If the number of elements removed exceeds k, return
            // the remaining number of unique elements
            if (elementsRemoved > k) {
                return frequencies.size() - i;
            }
        }

        // We have removed all elements, so no unique integers remain
        // Return 0 in this case
        return 0;
    }


    // V1-2
    // IDEA: MIN HEAP
    // https://leetcode.com/problems/least-number-of-unique-integers-after-k-removals/
    public int findLeastNumOfUniqueInts_1_2(int[] arr, int k) {
        // Map to track the frequencies of elements
        Map<Integer, Integer> map = new HashMap<>();
        for (int i : arr) {
            map.put(i, map.getOrDefault(i, 0) + 1);
        }

        // Min heap to track all the frequencies
        PriorityQueue<Integer> frequencies = new PriorityQueue<>(map.values());

        // Tracking the number of elements removed
        int elementsRemoved = 0;

        // Traversing all frequencies
        while (!frequencies.isEmpty()) {
            // Removing the least frequent element
            elementsRemoved += frequencies.peek();

            // If the number of elements removed exceeds k, return
            // the remaining number of unique elements
            if (elementsRemoved > k) {
                return frequencies.size();
            }
            frequencies.poll();
        }

        // We have removed all elements, so no unique integers remain
        // Return 0 in this case
        return 0;
    }


    // V1-3
    // IDEA: Counting Sort
    // https://leetcode.com/problems/least-number-of-unique-integers-after-k-removals/
    public int findLeastNumOfUniqueInts_1_3(int[] arr, int k) {
        // Map to track the frequencies of elements
        Map<Integer, Integer> map = new HashMap<>();
        for (int i : arr) {
            map.put(i, map.getOrDefault(i, 0) + 1);
        }

        int n = arr.length;

        // Array to track the frequencies of frequencies
        // The maximum possible frequency of any element
        // will be n, so we'll initialize this array with size n + 1
        int[] countOfFrequencies = new int[n + 1];

        // Populating countOfFrequencies array
        for (int count : map.values()) {
            countOfFrequencies[count]++;
        }

        // Variable to track the remaining number of unique elements
        int remainingUniqueElements = map.size();

        // Traversing over all possible frequencies
        for (int i = 1; i <= n; i++) {
            // For each possible frequency i, we'd like to remove as
            // many elements with that frequency as possible.
            // k / i represents the number of maximum possible elements
            // we could remove with k elements left to remove.
            // countOfFrequencies[i] represents the actual number of elements
            // with frequency i.
            int numElementsToRemove = Math.min(k / i, countOfFrequencies[i]);

            // Removing maximum possible elements
            k -= (i * numElementsToRemove);

            // numElementsToRemove is the count of unique elements removed
            remainingUniqueElements -= numElementsToRemove;

            // If the number of elements that can be removed is less
            // than the current frequency, we won't be able to remove
            // any more elements with a higher frequency so we can return
            // the remaining number of unique elements
            if (k < i) {
                return remainingUniqueElements;
            }
        }

        // We have traversed all possible frequencies i.e.,
        // removed all elements. Returning 0 in this case.
        return 0;
    }


    // V2
    // https://leetcode.ca/2019-12-20-1481-Least-Number-of-Unique-Integers-after-K-Removals/
    public int findLeastNumOfUniqueInts_2(int[] arr, int k) {
        Map<Integer, Integer> cnt = new HashMap<>();
        for (int x : arr) {
            cnt.merge(x, 1, Integer::sum);
        }
        List<Integer> nums = new ArrayList<>(cnt.values());
        Collections.sort(nums);
        for (int i = 0, m = nums.size(); i < m; ++i) {
            k -= nums.get(i);
            if (k < 0) {
                return m - i;
            }
        }
        return 0;
    }


}
